File: CS50's_Introduction_to_CS/Week1/Problem_set_1/credit.c
//
// Created by saksham on 16-06-2021.
//

#include <stdio.h>
#include <cs50.h>
#include <math.h>


int main(void)
{
    // Declaring variables
    int count = 0, arr[17], pos = 0;
    long rev = 0, remainder;
    long number = get_long("Number: ");
    long temp = number;
    int size = trunc(log10(number)) + 1;

    // Analysing size of the number:
    if (size < 13)
    {
        printf("INVALID\n");
    }
    else
    {
        while (number != 0)
        {
            remainder = number % 10;
            rev = rev * 10 + remainder;
            arr[pos] = remainder;
            pos++;
            number /= 10;
        }
        int sum = 0;
        for (int i = 1; i < pos; i += 2)
        {
            arr[i] *= 2;
            if (arr[i] <= 9)
            {
                sum += arr[i];
            }
            else
            {
                int waste = arr[i];
                while (waste != 0)
                {
                    sum += waste % 10;
                    waste /= 10;
                }
            }
        }
        for (int i = 0; i < pos; i += 2)
        {
            sum += arr[i];
        }
        if (sum % 10 == 0)
        {
            int startNum = 0;
            // Classifying cards:
            for (int i = 0; i < 2; i++)
            {
                startNum = startNum * 10 + (rev % 10);
                rev /= 10;
            }
            if ((startNum == 34 || startNum == 37) && pos == 15)
            {
                printf("AMEX\n");
            }
            else if (startNum >= 51 && startNum <= 55 && pos == 16)
            {
                printf("MASTERCARD\n");
            }
            else if (startNum / 10 == 4 && (pos == 13 || pos == 16))
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }

    }

}

File: CS50's_Introduction_to_CS/Week6/Problem_set_6/credit/credit.py
from cs50 import get_int

while True:
    # Getting value from user:
    number = get_int("Number: ")
    if int(number) > 0:
        break

# Declaring variables
number = str(number)
normal = number[::-1]

# Storing number in an array:
arr = [num for num in str(number)]
arr2 = []
for i in range(1, len(normal), 2):
    arr2.append(normal[i])
    arr.remove(normal[i])

# Validating length of the number
if len(number) < 13:
    print("INVALID")
else:
    temp = []
    # Multiplying digits starting from second-last by 2:
    for i in range(0, len(arr2)):
        newval = int(arr2[i]) * 2
        newval = str(newval)
        if len(newval) < 2:
            temp.append(newval)
        else:
            for digit in newval:
                temp.append(digit)
    arr2 = temp
    total = 0
    # Summation of the digit of products:
    for i in arr:
        total += int(i)
    # Summation of rest of the digits of the number:
    for j in arr2:
        total += int(j)
    # Checking whether the number is valid or not:
    if str(total)[-1] == "0":
        startNum = int(number[0:2])
        if (startNum == 34 or startNum == 37) and len(number) == 15:
            print("AMEX")
        elif 51 <= startNum <= 55 and len(number) == 16:
            print("MASTERCARD")
        elif startNum // 10 == 4 and (len(number) == 13 or len(number) == 16):
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")
